canary/S3ObjectTransport.cpp

```diff
@@ -0,0 +1,198 @@
+#include <aws/io/stream.h>
+
+using namespace Aws::Crt;
+
+S3ObjectTransport::S3ObjectTransport(
+    const Aws::Crt::String &region,
+    const Aws::Crt::String &bucket,
+    Aws::Crt::Io::TlsContext &tlsContext,
+    Aws::Crt::Io::ClientBootstrap &clientBootstrap,
+    const std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> &credsProvider,
+    const std::shared_ptr<Aws::Crt::Auth::Sigv4HttpRequestSigner> &signer,
+    size_t maxCons)
+    : m_signer(signer), m_credsProvider(credsProvider), m_region(region), m_bucketName(bucket)
+{
+    Http::HttpClientConnectionManagerOptions connectionManagerOptions;
+    m_endpoint = m_bucketName + ".s3." + m_region + ".amazonaws.com";
+    connectionManagerOptions.ConnectionOptions.HostName = m_endpoint;
+    connectionManagerOptions.ConnectionOptions.Port = 443;
+    connectionManagerOptions.ConnectionOptions.SocketOptions.SetConnectTimeoutMs(3000);
+    connectionManagerOptions.ConnectionOptions.SocketOptions.SetSocketType(AWS_SOCKET_STREAM);
+    connectionManagerOptions.ConnectionOptions.InitialWindowSize = SIZE_MAX;
+
+    aws_byte_cursor serverName = ByteCursorFromCString(m_endpoint.c_str());
+
+    auto connOptions = tlsContext.NewConnectionOptions();
+    connOptions.SetServerName(serverName);
+    connectionManagerOptions.ConnectionOptions.TlsOptions = connOptions;
+    connectionManagerOptions.ConnectionOptions.Bootstrap = &clientBootstrap;
+    connectionManagerOptions.MaxConnections = maxCons;
+
+    m_connManager =
+        Http::HttpClientConnectionManager::NewClientConnectionManager(connectionManagerOptions, g_allocator);
+
+    m_hostHeader.name = ByteCursorFromCString("host");
+    m_hostHeader.value = ByteCursorFromCString(m_endpoint.c_str());
+
+    m_contentTypeHeader.name = ByteCursorFromCString("content-type");
+    m_contentTypeHeader.value = ByteCursorFromCString("text/plain");
+}
+
+void S3ObjectTransport::PutObject(
+    const Aws::Crt::String &key,
+    struct aws_input_stream *inputStream,
+    TransportOpCompleted transportOpCompleted)
+{
+    auto request = MakeShared<Http::HttpRequest>(g_allocator, g_allocator);
+    request->AddHeader(m_hostHeader);
+    request->AddHeader(m_contentTypeHeader);
+
+    Http::HttpHeader contentLength;
+    contentLength.name = ByteCursorFromCString("content-length");
+
+    int64_t streamLen = 0;
+
+    aws_input_stream_get_length(inputStream, &streamLen);
+    StringStream intValue;
+    intValue << streamLen;
+    String contentLengthVal = intValue.str();
+    contentLength.value = ByteCursorFromCString(contentLengthVal.c_str());
+    request->AddHeader(contentLength);
+
+    aws_http_message_set_body_stream(request->GetUnderlyingMessage(), inputStream);
+    request->SetMethod(aws_http_method_put);
+
+    String keyPath = "/" + key;
+    ByteCursor path = ByteCursorFromCString(keyPath.c_str());
+    request->SetPath(path);
+
+    auto signingConfig = MakeShared<Auth::AwsSigningConfig>(g_allocator, g_allocator);
```

Comment:
This doesn't need to outlive the call to sign_request anymore, so you can just put it on the stack and not worry about lifetimes.